libsql-sqlite3/src/wal.c

```diff
@@ -4250,150 +4151,260 @@ static void libsqlGetWalPathname(char *buf, const char *orig, int orig_len) {
+    RefCountedWalManager *p = (RefCountedWalManager*)sqlite3MallocZero(sizeof(RefCountedWalManager));
+    if (!p) return SQLITE_NOMEM;
+    p->n = 1;
+    p->ref = wal_manager;
+    p->is_static = 0;
+    *out = p;
+    return SQLITE_OK;
+}
 
-  while (1) {
-    if (m == NULL) {
-      prev->pNext = pWalMethods;
-      pWalMethods->pNext = NULL;
-      break;
-    }
-    if (strcmp(m->zName, pWalMethods->zName) == 0) {
-      prev->pNext = pWalMethods;
-      pWalMethods->pNext = m->pNext;
-      break;
+/*
+ * Decrease the ref count and call the wal_manager destructor when the count reaches 0.
+ * Must be called from withing a critical section.
+ */
+void destroy_wal_manager(RefCountedWalManager *p) {
+    if (p->is_static) return;
+    assert(p->n != 0);
+    p->n -= 1;
+    if (p->n == 0) {
+        (p->ref.xDestroy)(p->ref.pData);
+        sqlite3_free(p);
     }
-    prev = m;
-    m = m->pNext;
-  }
-  sqlite3_mutex_leave(mutex);
-  return SQLITE_OK;
 }
 
-int libsql_wal_methods_unregister(libsql_wal_methods *pWalMethods) {
-  if (strncmp(pWalMethods->zName, "default", 7) == 0) {
-    return SQLITE_MISUSE;
-  }
-  MUTEX_LOGIC(sqlite3_mutex *mutex;)
-  MUTEX_LOGIC( mutex = sqlite3MutexAlloc(SQLITE_MUTEX_STATIC_MAIN); )
-  sqlite3_mutex_enter(mutex);
+/*
+ * Increase the ref count and return a pointer to the wal.
+ * Must be called from withing a critical section.
+ * Return NULL if the passed pointer ref count is already 0.
+ */
+RefCountedWalManager* clone_wal_manager(RefCountedWalManager *p) {
+    assert(p->n != 0);
+    p->n += 1;
+    return p;
+}
 
-  libsql_wal_methods *prev = libsql_wal_methods_find("default");
-  libsql_wal_methods *m = prev->pNext;
+const libsql_wal_manager sqlite3_wal_manager = { 
+    .pData = NULL, 
+    .xOpen = (int (*)(wal_manager_impl *, sqlite3_vfs *, sqlite3_file *, int, long long, const char*, libsql_wal *))sqlite3WalOpen,
+    .xClose = (int (*)(wal_manager_impl *, wal_impl *, sqlite3 *, int, int, unsigned char *))sqlite3WalClose,
+    .bUsesShm = 1,
+    .xLogDestroy = (int (*)(wal_manager_impl *, sqlite3_vfs*, const char*))sqlite3LogDestroy,
+    .xLogExists = (int (*)(wal_manager_impl *, sqlite3_vfs*, const char*, int *))sqlite3LogExists,
+    .xDestroy =(void (*)(wal_manager_impl*))sqlite3DestroyWalManager,
+};
 
-  while (m != NULL) {
-    if (strcmp(m->zName, pWalMethods->zName) == 0) {
-      prev->pNext = m->pNext;
-      break;
+void make_sqlite3_wal_manager_rc(RefCountedWalManager **out) {
+    static int initialized = 0;
+    static RefCountedWalManager manager = { 0 };
+    /*
+     * re-initializing is idempotent
+     */
+    if (!initialized) {
+        manager.is_static = 1;
+        manager.ref = sqlite3_wal_manager;
+        manager.n = 1;
+        initialized = 1;
     }
-    prev = m;
-    m = m->pNext;
-  }
-  sqlite3_mutex_leave(mutex);
-  return SQLITE_OK;
-}
 
-libsql_wal_methods *libsql_wal_methods_next(libsql_wal_methods *w) {
-  return w->pNext;
+    *out = &manager;
```

Comment:
Would that compile? Then we avoid the `initialized` dance, static are initialized lazily anyway

```
RefCountedWalManager *make_sqlite3_wal_manager_rc() {
    static RefCountedWalManager manager = {
        .is_static = 1,
        .ref = sqlite3_wal_manager,
        .n = 1
    };
    return &manager;
}
```

Reply:
no that doesn't compile because because somehow C is unable to compute `.ref = sqlite3_wal_manager,` as a static expression 🤦

Reply:
that was my initial attempt btw